ksqldb-engine/src/main/java/io/confluent/ksql/internal/ThroughputMetricsReporter.java

```diff
@@ -0,0 +1,259 @@
+
+    // CHECKSTYLE_RULES.OFF: LineLength
+    private static final Map<String, Map<String, Map<MetricName, ThroughputTotalMetric>>> registeredMetrics = new HashMap<>();
+    // CHECKSTYLE_RULES.ON: LineLength
+    private static final Map<String, String> customTags = new HashMap<>();
+    private Metrics metricRegistry;
+
+    @Override
+    public void init(final List<KafkaMetric> initial) {
+    }
+
+    @Override
+    public synchronized void configure(final Map<String, ?> configMap) {
+        this.metricRegistry = (Metrics) requireNonNull(
+            configMap.get(KsqlConfig.KSQL_INTERNAL_METRICS_CONFIG)
+        );
+        customTags.putAll(KsqlConfig.getStringAsMap(KsqlConfig.KSQL_CUSTOM_METRICS_TAGS, configMap));
+    }
+
+    @Override
+    public void metricChange(final KafkaMetric metric) {
+        if (!THROUGHPUT_METRIC_NAMES.contains(metric.metricName().name())) {
+            return;
+        }
+        addMetric(
+            metric,
+            getQueryId(metric),
+            getTopic(metric)
+        );
+    }
+
+    @Override
+    public void metricRemoval(final KafkaMetric metric) {
+        if (!THROUGHPUT_METRIC_NAMES.contains(metric.metricName().name())) {
+            return;
+        }
+
+        removeMetric(
+            metric,
+            getQueryId(metric),
+            getTopic(metric)
+        );
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public Set<String> reconfigurableConfigs() {
+        return null;
+    }
+
+    @Override
+    public void validateReconfiguration(final Map<String, ?> configs) throws ConfigException {
+
+    }
+
+    @Override
+    public void reconfigure(final Map<String, ?> configs) {
+
+    }
+
+    @Override
+    public void contextChange(final MetricsContext metricsContext) {
+
+    }
+
+    private synchronized void addMetric(
+        final KafkaMetric metric,
+        final String queryId,
+        final String topic
+    ) {
+        final MetricName metricName = getMetricNameWithQueryIdTag(queryId, metric.metricName());
+        final Map<String, String> queryTopicThroughputMetricTags =
+            getQueryMetricTags(queryId, metric.metricName().tags());
+
+        LOGGER.debug("Adding {} metric for query={} and topic={}", metricName.name(), queryId, topic);
+
+        if (!registeredMetrics.containsKey(queryId)) {
+            registeredMetrics.put(queryId, new HashMap<>());
+        }
+        if (!registeredMetrics.get(queryId).containsKey(topic)) {
+            registeredMetrics.get(queryId).put(topic, new HashMap<>());
+        }
+
+        final ThroughputTotalMetric newMetric;
+        if (registeredMetrics.get(queryId).get(topic).containsKey(metricName)) {
+            newMetric = registeredMetrics.get(queryId).get(topic).get(metricName);
+            newMetric.add(metricName, metric);
+        } else {
+            newMetric = registeredMetrics.get(queryId).get(topic).get(metricName);
```

Comment:
Assuming it is indeed a typo, then I'm not sure I understand this logic here: it seems the goal is to use a single `ThroughputTotalMetric` per-query per-topic, with differentiated `bytes` and `records`, to aggregate all metrics across different task-id / processor-node-id / thread-id, right? But from the code it seems not achieving this goal since each metricName (with the whole tag map) would be getting its own `ThroughputTotalMetric` and hence the inner map of the `ThroughputTotalMetric` would always contain a single element.

Reply:
Plus, I'm wondering if it's a bit over-complicated to achieve this logic plus being able to remove the `ThroughputTotalMetric ` metric when all of its associated finer grained metrics are gone?

Is it efficient to just maintain a private static final `Map<String, Map<String, Map<String, ThroughputTotalMetric>>> registeredMetrics` where the inner-most map's key are just `bytes` and `records`, since the `ThroughputTotalMetric` itself maintains the map of its associated finer-grained metrics. And then upon `metricRemoval` we can just check if the `ThroughputTotalMetric`'s inner map is empty and if yes remove that metric from the `registeredMetrics`?

Reply:
> Should this be newMetric = new ThroughputTotalMetric...

haha whoops -- yes that was definitely a typo, thanks.

> each metricName (with the whole tag map) would be getting its own ThroughputTotalMetric and hence the inner map of the ThroughputTotalMetric would always contain a single element.

Ah ok I think something I changed at the last minute did indeed introduce a bug and is what's causing this confusion -- the innermost map in `registeredMetrics` was supposed to be keyed by the `metricName.name()`, eg `bytes-consumed-total`, `records-produced-total`, etc -- I tried to simplify things and make the code easier to understand by changing this to the plain `MetricName` but in fact that's not the same thing as there may be multiple metrics for a given metric-type & topic &  the reason corresponding to different `taskId` and `processorNodeId` values in the tags. So yes, the way it is now would indeed result in just a single metric per `ThroughputTotalMetric` resulting in non-aggregated metrics. I'll change it back to the way it was, ie keyed by `metricName.name()`, which I think was what you were getting at with this suggestion? 

>  just maintain a private static final Map<String, Map<String, Map<String, ThroughputTotalMetric>>> registeredMetrics where the inner-most map's key are just bytes and records

However the second part of your suggestion I'm not quite seeing, are you suggesting we don't check & clean up the outer maps when we've removed the last ThroughputMetric in the topic map, or the last topic entry in the query map? It may be a bit longwinded but I do think we need to maintain these maps or else we'll leak memory over time as queries come and go

> And then upon metricRemoval we can just check if the ThroughputTotalMetric's inner map is empty and if yes remove that metric from the registeredMetrics?

Reply:
Ok I'm actually going to leave the innermost map key type as a `MetricName` rather than `String: MetricName.name()` just because I think it helps to add clarity around the meaning of each String/key -- I just need to fix the `MetricName` to remove the `taskId` and `processorNodeId` tags, then we should correctly end up adding and aggregating over each of these metrics for a given queryId-topic-metricType combination in the `ThroughputTotalMetric`. 

Lmk if that makes sense or if you still have questions -- can also hop on a quick call tomorrow to go over this stuff if there's anything left to resolve